lib/Service/ContactsIntegration.php

```diff
@@ -118,10 +118,14 @@ public function getMatchingRecipient(string $userId, string $term): array {
 					continue;
 				}
 				$lowerTerm = strtolower($term);
-				if (!$isSystemUser || $isInSameGroup || ($lowerTerm !== '' && (
-					($shareeEnumerationFullMatch && !empty($fn) && $lowerTerm === strtolower($fn)) ||
-					($shareeEnumerationFullMatchUserId && $lowerTerm === strtolower($id)) ||
-					($shareeEnumerationFullMatchEmail && $lowerTerm === strtolower($e))))) {
+				if (!$isSystemUser || // Not a user in SAB
```

Comment:
similar to the first if in the loop we could flip the conditions and change to a series of small checks with continue

the code is indeed very hard to read. it was also hard to read before :) 

Add a test to https://github.com/nextcloud/mail/blob/main/tests/Unit/Service/ContactsIntegrationTest.php too for this case, please